### PR TITLE
Detect mismatches in begin and end tokens returned by JSON tokenizer FST

### PR DESCRIPTION
**Pull request by shrshi**
## Description
Addresses #15820 

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/json/json_tree.cu [hunk 1]
```diff
@@ -272,6 +272,63 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
...
+                   "Some begin token does not have a matching end token");
+    } else {
+      auto not_ok = thrust::transform_reduce(
+        rmm::exec_policy(stream),
+        tokens.begin(),
+        tokens.end(),
+        cuda::proclaim_return_type<std::uint8_t>([] __device__(auto token) -> std::uint8_t {
+          std::uint8_t token_bits{0};
+          switch (token) {
+            case token_t::StructBegin: [[fallthrough]];
```
Comment (vuule) [inline cpp/src/io/json/json_tree.cu:293]: No need for `[[fallthrough]]` when the case is empty. I misunderstood that guideline at first so we have a lot of unnecessary `[[fallthrough]]`s in the code :grimacing:

File: cpp/src/io/json/json_tree.cu [hunk 2]
```diff
@@ -272,6 +272,63 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
...
+    // Struct, List, StructMember, FieldName, String, Value
+    // Corner case: If the input is a literal such as "true" or "false" that does not end in a
+    // newline character, then we will have only the ValueBegin token without the corresponding
+    // ValueEnd
+    if (tokens.size() == 1) {
+      auto h_tokens = cudf::detail::make_std_vector_sync(tokens, stream);
+      CUDF_EXPECTS(h_tokens[0] == token_t::ValueBegin,
+                   "Some begin token does not have a matching end token");
+    } else {
+      auto not_ok = thrust::transform_reduce(
```
Comment (vuule) [inline cpp/src/io/json/json_tree.cu:286]: does this work when we have two (same) begin tokens without any end tokens?

**Comment by shrshi** [comment]
When the last JSONL record in the input is incomplete, and error recovery is set to FAIL, then an error token is not emitted after the entire input is read, which gives rise to above bug. Even though a newline character is [inserted](https://github.com/rapidsai/cudf/blob/76b35adec49d85cf23d4a32a44588c856234f140/cpp/src/io/json/read_json.cu#L271) after the last record, error tokens are emitted on [line break](https://github.com/rapidsai/cudf/blob/76b35adec49d85cf23d4a32a44588c856234f140/cpp/src/io/json/nested_json_gpu.cu#L894) only when `recover_with_null` is enabled. Instead of modifying the translation table for the PDA, a simpler approach to detecting if the last line in incomplete is to inspect the logical stack after the bracket-brace FST. For valid inputs, the stack should be empty so the incomplete last line error can be captured.

File: cpp/src/io/json/json_tree.cu (hunk 2 above)
Comment (shrshi) [reply cpp/src/io/json/json_tree.cu]: This logic has been replaced a check on the logical stack - if the logical stack is non-empty after the bracket-brace FST then the last record is incomplete.

**Review by vuule — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

**Pull request merged**